impl/meson.build

```diff
@@ -47,30 +63,60 @@ targetLibraryBuildable = not (meson.is_cross_build() and (
 	or target_machine.cpu_family() == 'avr'
 ))
 
+library_type = get_option('default_library')
+
+if library_type != 'static'
+	libSubstrateArgs = []
+else
+	libSubstrateArgs = ['-DSUBSTRATE_BUILD_STATIC']
+endif
+
 if targetLibraryBuildable
-	libSubstrate = library(
-		'substrate',
-		libSubstrateSrcs,
-		c_args: ['-DSUBSTRATE_BUILD_INTERNAL'],
-		cpp_args: ['-DSUBSTRATE_BUILD_INTERNAL'],
-		include_directories: include_directories('..'),
-		dependencies: deps,
-		gnu_symbol_visibility: 'inlineshidden',
-		implicit_include_directories: false,
-		pic: true,
-		install: (not meson.is_subproject()),
-	)
+	if library_type != 'static'
+		libSubstrateShared = shared_library(
+			'substrate',
+			libSubstrateSrcs,
+			c_args: ['-DSUBSTRATE_BUILD_INTERNAL'],
+			cpp_args: ['-DSUBSTRATE_BUILD_INTERNAL'],
+			include_directories: include_directories('..'),
+			dependencies: deps,
+			gnu_symbol_visibility: 'inlineshidden',
+			implicit_include_directories: false,
+			install: (not meson.is_subproject()),
+		)
+	endif
+
+	if library_type != 'shared'
+		libSubstrateStatic = static_library(
+			'substrateStatic',
+			libSubstrateSrcs,
+			c_args: ['-DSUBSTRATE_BUILD_INTERNAL'] + libSubstrateArgs,
+			cpp_args: ['-DSUBSTRATE_BUILD_INTERNAL'] + libSubstrateArgs,
+			include_directories: include_directories('..'),
+			dependencies: deps,
+			gnu_symbol_visibility: 'inlineshidden',
+			implicit_include_directories: false,
+			pic: true,
+			install: (not meson.is_subproject()),
+		)
+	endif
+
+	if library_type != 'static'
+		libSubstrate = libSubstrateShared
+	else
+		libSubstrate = libSubstrateStatic
+	endif
 else
 	libSubstrate = []
 endif
 
 # In a cross-build scenario, always build a version of the library for the host too.
 if meson.is_cross_build()
-	libSubstrateNative = library(
```

Comment:
Is there a particular reason here why we're now limiting this to static-only for the host platform ("native") build?

Reply:
Same as with the above-- if the `default_library` is specified as `both`, Meson provides no way to ensure that the native version will have the correct exports macro. One will be (correctly) get `__declspec(dllexport)`, and the other will be a dud that'll fail at link time.

Reply:
Hmm.. there are instances (SPIFlashProgrammer) when we intended to link substrate as a shared library.. but given that the question is what to do for the `both` case.. 🤔 
Would it be feasible to make `default_library=static` generate a static library and everything else only generate a shared? If you think it would be too much work then we'll accept this as-is as it is a change in behaviour but it's not disastrous.

Reply:
Just did it, I've gotten the reply I expected from xclaesse; Meson doesn't yet have working support for `default_library = 'both'` on Windows.